feat(dapps)_: extend and improve signing

stefandunca opened the pull request:
### Updates: [#15361](https://github.com/status-im/status-desktop/issues/15361)

Required by https://github.com/status-im/status-desktop/pull/15424

Add `wallet_SafeSignTypedDataForDApps` with support for `eth_signTypedData` and `eth_signTypedData_v4`
Reject if the chain to sign doesn't matches the target chain for typed data signing

Add `wallet_SignMessageForDApps` with support for signing messages in a safe way as per EIP-191 v45

Add `wallet_HashMessageForSigning` to hash messages for signing on the client side (keycard)

Deprecate `wallet_SignTypedDataV4``

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (23)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | e3901ed1 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/1/) | 2024-07-02 10:46:51 | ~1 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/1/consoleText) |
| :x: | e3901ed1 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/1/) | 2024-07-02 10:47:01 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/1/consoleText) |
| :x: | e3901ed1 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/1/) | 2024-07-02 10:47:07 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/1/consoleText) |
| :heavy_multiplication_x: | e3901ed1 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/1/) | 2024-07-02 10:48:02 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/1//consoleText) |
| | | | | | | |
| :heavy_multiplication_x: | 8585ec1c | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/2/) | 2024-07-02 21:48:39 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/2//consoleText) |
| :heavy_check_mark: | 8585ec1c | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/2/) | 2024-07-02 21:50:55 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240702-214709-8585ec-pr5463.zip) |
| :heavy_check_mark: | 8585ec1c | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/2/) | 2024-07-02 21:51:06 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240702-214709-8585ec-pr5463.zip) |
| :heavy_check_mark: | 8585ec1c | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/2/) | 2024-07-02 21:51:51 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240702-214709-8585ec-pr5463.aar) |
| | | | | | | |
| :heavy_check_mark: | 0ef08d3e | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/3/) | 2024-07-03 16:10:36 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240703-160529-0ef08d-pr5463.zip) |
| | | | | | | |
| :heavy_check_mark: | c116fba6 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/4/) | 2024-07-03 16:11:48 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240703-160925-c116fb-pr5463.aar) |
| :heavy_check_mark: | c116fba6 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/4/) | 2024-07-03 16:11:56 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240703-160923-c116fb-pr5463.zip) |
| :heavy_check_mark: | c116fba6 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/4/) | 2024-07-03 16:14:20 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240703-161055-c116fb-pr5463.zip) |
| :heavy_multiplication_x: | c116fba6 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/4/) | 2024-07-03 16:53:53 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/4//consoleText) |
| | | | | | | |
| :heavy_check_mark: | d2ec9377 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/5/) | 2024-07-05 11:18:26 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240705-111545-d2ec93-pr5463.aar) |
| :heavy_check_mark: | d2ec9377 | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5463/1/) | 2024-07-05 11:18:30 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5463/1//consoleText) |
| :heavy_check_mark: | d2ec9377 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/5/) | 2024-07-05 11:18:36 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240705-111549-d2ec93-pr5463.zip) |
| :heavy_check_mark: | d2ec9377 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/5/) | 2024-07-05 11:19:00 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240705-111549-d2ec93-pr5463.zip) |
| :heavy_check_mark: | d2ec9377 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/5/) | 2024-07-05 11:59:53 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/5//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 7ece1b9b | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/6/) | 2024-07-05 13:52:26 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240705-135008-7ece1b-pr5463.aar) |
| :heavy_check_mark: | 7ece1b9b | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5463/2/) | 2024-07-05 13:52:58 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5463/2//consoleText) |
| :heavy_check_mark: | 7ece1b9b | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/6/) | 2024-07-05 13:53:11 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240705-135008-7ece1b-pr5463.zip) |
| :heavy_check_mark: | 7ece1b9b | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/6/) | 2024-07-05 13:54:04 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240705-135012-7ece1b-pr5463.zip) |
| :heavy_check_mark: | 7ece1b9b | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/6/) | 2024-07-05 14:33:43 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/6//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | b03e09bb | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/7/) | 2024-07-08 17:54:46 | ~1 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/7//consoleText) |
| :heavy_check_mark: | b03e09bb | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/7/) | 2024-07-08 17:55:30 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240708-175311-b03e09-pr5463.aar) |
| :heavy_check_mark: | b03e09bb | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/7/) | 2024-07-08 17:55:38 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240708-175316-b03e09-pr5463.zip) |
| :heavy_check_mark: | b03e09bb | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5463/3/) | 2024-07-08 17:55:48 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5463/3//consoleText) |
| :heavy_check_mark: | b03e09bb | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/7/) | 2024-07-08 17:55:55 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240708-175311-b03e09-pr5463.zip) |
| | | | | | | |
| :heavy_check_mark: | 62297483 | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5463/8/) | 2024-07-08 18:39:52 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240708-183727-622974-pr5463.zip) |
| :heavy_check_mark: | 62297483 | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5463/8/) | 2024-07-08 18:41:02 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240708-183731-622974-pr5463.zip) |
| :heavy_check_mark: | 62297483 | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5463/8/) | 2024-07-08 18:42:40 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240708-183727-622974-pr5463.aar) |
| :heavy_check_mark: | 62297483 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/8/) | 2024-07-08 19:20:09 | ~42 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5463/8//consoleText) |

stefandunca committed:
feat(dapps)_: extend and improve sign

Add `wallet_SafeSignTypedDataForDApps` with support for
`eth_signTypedData` and `eth_signTypedData_v4`
Reject if the chain to sign doesn't matches the target chain
for typed data signing

Add `wallet_HashMessageForSigning` with to support hashing messages
for signing in a safe way as per EIP-191 v45 and supporting
to hash messages for signing on the client side (keycard)

Deprecate `wallet_SignTypedDataV4``

Updates: #15361

stefandunca commented on services/wallet/api.go:
```diff
@@ -853,8 +854,37 @@ func (api *API) GetWalletConnectDapps(ctx context.Context, validAtTimestamp int6
+// SignTypedDataV4 dApps use it to execute "eth_signTypedData_v4" requests
+// the formatted typed data will be prefixed with \x19\x01 based on the EIP-712
+// @deprecated
 func (api *API) SignTypedDataV4(typedJson string, address string, password string) (types.HexBytes, error) {
```
The usage of this API should be replaced by `SafeSignTypedDataForDApps`

saledjenic commented on services/wallet/api.go:
```diff
@@ -853,8 +854,37 @@ func (api *API) GetWalletConnectDapps(ctx context.Context, validAtTimestamp int6
+}
+
+// SignMessageForDApps is used for handling dApps requests for "personal_sign" and "eth_sign" requests
+func (api *API) SignMessageForDApps(ctx context.Context, message types.HexBytes, address common.Address, password string) (string, error) {
```
I am not a fan of naming functions like `SignMessageForDApps`, is it matter if the signing is for dapps or not.

Since we already have `SignMessage` and `HashMessageForSigning` functions exposed by this API you can use them on the client side to achieve the same or if you really want to have those two calls combined in this API you can name it `HashMessageAndSign`.

stefandunca replied:
Good point, I forgot that there is already a `wallet_SignMessage`; I will use that. Thanks.

clauxx replied:
what about just having smth like SignMessageEIP191, which makes it clearer how it's different from SignMessage.

stefandunca replied:
I'm now testing a mix of both recommendations use `HashMessageEIP191` for hashing and the existing `SignMessage` for signing

saledjenic commented on services/wallet/api.go:
```diff
@@ -874,6 +887,48 @@ func (api *API) SignTypedDataV4(typedJson string, address string, password strin
+		return types.HexBytes{}, err
+	}
+
+	var typed typeddata.TypedData
```
This, lines `903` - `907` could go under "then"block of `if legacy {`.
@stefandunca

stefandunca replied:
The un-marshaing and following call to `ValidateChainID` is used in both branches of legacy call. I added a comment to raise the awareness that
